circleguard/investigator.py

```diff
@@ -206,11 +206,153 @@ def median_frametime(replay):
 
         Notes
         -----
-        Median is used instead of mean to lessen the effect of outliers.
+        Frametimes on keypresses are added to the next frametime to make frametimes more accurate.
+        Frametimes are further adjusted for random low frametimes that appear.
         """
         # replay.t is cumsum so convert it back to "time since previous frame"
         t = np.diff(replay.t)
-        return np.median(t)
+
+        # filters out insignificant frametimes
+        unique, counts = np.unique(t, return_counts=True)
+        unique = unique[counts > 4]
+        counts = counts[counts > 4]
+
+        # placeholder 0, so t matches k
+        t = np.concatenate(([0], t))
+        replay.k = replay.k[np.isin(t, unique)]
+        t = t[np.isin(t, unique)]
+
+        def mode(x):
+            if len(x) == 0:
+                return 0
+            values, counts = np.unique(x, return_counts=True)
+            m = counts.argmax()
+            return values[m]
+
+        filtered_frametimes = []
+        previous_key = 0
+        combine_next = False
+        # for very fast buzz sliders
+        count_lowframes = 0
+        current_mode = 5
+        for i in range(len(t)):
+            # update mode only for first 100 for efficiency
+            if i < 100:
+                current_mode = max(5, mode(filtered_frametimes))
+            else:
+                # unless mode is still less than 5
+                if current_mode < 5:
+                    current_mode = max(5, mode(filtered_frametimes))
+            if t[i] < current_mode - 5:
+                count_lowframes += 1
+            else:
+                if count_lowframes > 8:
+                    filtered_frametimes = filtered_frametimes[:len(filtered_frametimes) - count_lowframes]
+                count_lowframes = 0
+            k = replay.k[i]
+            if previous_key == replay.k[i]:
+                if combine_next == False:
+                    filtered_frametimes.append(t[i])
+                else:
+                    # adds times together if they don't go over the mode too much
+                    combined_time = t[i-1] + t[i]
+                    if combined_time > current_mode + 5:
+                        filtered_frametimes.extend([t[i-1], t[i]])
+                    else:
+                        filtered_frametimes.append(combined_time)
+                combine_next = False
+            else:
+                combine_next = True
+            previous_key = k
+        filtered_frametimes = filtered_frametimes[1:]
```

Comment:
I'm confused as to what this entire thing is doing. Why does mode start at 5? What are these magic numbers doing here (`combined_time > current_mode + 5`, `count_lowframes > 8`)? Please extract them out as constants.

There's also many indentation levels and little documentation. When something is this complicated, please extract it to a new method(s) with a docstring.